### PR TITLE
Support for Hyperspace 0.4.0

### PR DESCRIPTION
**Pull request by AFFogarty**
This PR adds full support for [Hyperspace 0.4.0](https://github.com/microsoft/hyperspace/releases/tag/v0.4.0).  This includes new APIs `optimizeIndex()` and `index()` introduced in Hyperspace 0.3.0 and 0.4.0, respectively.

### CODE REVIEW

File: src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs [hunk 1]
```diff
@@ -70,11 +70,37 @@ public Hyperspace(SparkSession spark)
...
-        /// Update indexes for the latest version of the data.
+        /// Update indexes for the latest version of the data. This API provides a few supported refresh
+        /// modes as listed below.
         /// </summary>
         /// <param name="indexName">Name of the index to refresh.</param>
+        /// <param name="mode">Refresh mode. Currently supported modes are <c>incremental</c> and
+        /// <c>full</c>.</param>
         [HyperspaceSince(HyperspaceVersions.V0_0_1)]
-        public void RefreshIndex(string indexName) => _jvmObject.Invoke("refreshIndex", indexName);
+        public void RefreshIndex(string indexName, string mode = "full") =>
```
Comment (AFFogarty) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs:80]: Do you think we should keep `mode` as a `string`, or introduce an `enum`?

Comment (imback82) [inline src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs:80]: Can we separate out the API that takes in the mode, which is from 0.3?

Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs:80]: Sure.  Do you think we should keep `mode` as a `string` in .Net, or make it an `enum`?

Comment (imback82) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs:80]: `string` is fine.

Comment (AFFogarty) [reply src/csharp/Extensions/Microsoft.Spark.Extensions.Hyperspace/Hyperspace.cs]: Updated the PR with `RefreshIndex` split in 2.

**Review by imback82 — APPROVED** [review]
LGTM (pending tests), thanks @AFFogarty!

**Pull request merged**